### PR TITLE
Select files and directories in a package to download

### PR DESCRIPTION
**Pull request by fiskus**
Backend: https://github.com/quiltdata/enterprise/pull/779

Screenshot from 2024-10-11 17-22-23
Screenshot from 2024-10-11 17-22-51

Also, some refactoring and fixing minor bugs:

* `SelectionItem { logicalKey: string }` type instead of a `string`
* Better aligning (fixing MUI bug)
* Re-use components

- [x] [Changelog](../tree/master/docs/CHANGELOG.md) entry (skip if change is not significant to end users, e.g. docs only)

### CODE REVIEW

File: catalog/app/containers/Bucket/Meta.tsx [hunk 1]
```diff
@@ -81,6 +81,7 @@ export function PackageMetaSection({ meta, preferences }: PackageMetaSectionProp
   const classes = usePackageMetaStyles()
   if (!meta || R.isEmpty(meta)) return null
   const { message, user_meta: userMeta, workflow } = meta
+  if (!message && !userMeta && !workflow) return null
```
Comment (nl0) [inline catalog/app/containers/Bucket/Meta.tsx:84]: i think you should remove the `R.isEmpty(meta)` check above if you're adding this

File: catalog/CHANGELOG.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ where verb is one of
 
 ## Changes
 
+- [Added] Select files and directories in packages to download ([#4173](https://github.com/quiltdata/quilt/pull/4173))
```
Comment (nl0) [inline catalog/CHANGELOG.md:20]:
```suggestion
- [Added] Selective package downloading ([#4173](https://github.com/quiltdata/quilt/pull/4173))
```
maybe sth like this?

**Review by nl0 — APPROVED** [review]
looks reasonable

File: catalog/CHANGELOG.md (hunk 2 above)
Comment (fiskus) [reply catalog/CHANGELOG.md:20]: I started with this, but tried to find more specific and simpler phrasing. However, using verbs in my variant is hanging in the air.

File: catalog/app/containers/Bucket/Meta.tsx (hunk 1 above)
Comment (fiskus) [reply catalog/app/containers/Bucket/Meta.tsx:84]: Oh, thanks

**Comment by fiskus** [comment]
Used `React.Provider` for passing selection data instead of `props`. It was required to use `utils/Dialog`, because it is initialized in callback and doesn't react to `props` changes.

Also, fixed route guarding: confirming navigation from package page when something is selected. The condition turned out to be quite sprawling: https://github.com/quiltdata/quilt/pull/4173/files#diff-95c902495a06ff31c5e3a2bede51af727048f9151bede5b5195a4f2cb8fc5090R106

**Review by nl0 — APPROVED** [review]

**Pull request merged**